### ISSUE TITLE
Make HTTP cache prune-able on the command line?

### ISSUE THREAD

**Issue by fritzmg** [contao/contao#7534]
Currently the HTTP cache of the Contao Managed Edition automatically prunes the cache after 5000 write operations. See for reference:

* https://github.com/contao/contao/blob/eb73b4ac735a9af2bf58206c371750a56c90d794/manager-bundle/src/HttpKernel/ContaoCache.php#L78
* https://github.com/Toflar/psr6-symfony-http-cache-store/blob/9b3eb5b5f449a886b3ce202ddf3705e3d2ac79c9/src/Psr6Store.php#L188-L189
* https://github.com/Toflar/psr6-symfony-http-cache-store/blob/9b3eb5b5f449a886b3ce202ddf3705e3d2ac79c9/src/Psr6Store.php#L446-L465

This happens within the HTTP request. This is fine for small to medium sized websites. However, we have one case where sometimes the amount of cache entries is so large that the pruning operation takes so long it runs into the `max_execution_time` limit.

Ideally what I would like to have instead is a command that I can execute e.g. daily that prunes the HTTP cache for me, instead of doing it during a request.

Primarily I wanted to discuss whether this makes sense in the first place and where to implement this (or another solution).

**Comment by Toflar** [comment]
Well, you have to think of the HTTP Cache as its separate instance. A command runs within the app, the cache does not.
So at the end of the day, it is an HTTP request anyway. No matter how you implement it 😊 
So integrating it with the Messenger or anything like that won't work.

**Comment by fritzmg** [comment]
I suppose so. My current workaround would be to patch the `prune_threshold` in the `ContaoCache` to `0` and then implement a command that creates a `Psr6Store` instance manually with the same settings as `ContaoCache` would do (sans `prune_threshold`) where I can then call its `->prune()` method.

But what would be the correct way to solve this?

**Comment by Toflar** [comment]
> But what would be the correct way to solve this?

I don't have any solution 🤷‍♂️

**Comment by Toflar** [comment]
Maybe...hmm. So the pruning takes so long because we use the `FilesystemAdapter` which consists of files and thus opens every file. For SQL this would be way faster because it's one DELETE query which can use an index.
So you might try to use the `PdoAdapter` together with SQLite for those systems that support it. We should check if performance is good for regular reads etc. as well but it might be even better than working with the filesystem as we use content digests. Content digests cause 2 lookups per read, so in the `FilesystemAdapter` that would be 2 `fopen()` handles. With SQLite this would end up with just one so it might be indeed faster.

**Comment by fritzmg** [comment]
> Maybe...hmm. So the pruning takes so long because we use the `FilesystemAdapter` which consists of files and thus opens every file.

Yes, at least I think so. It runs into the `max_execution_time` here: https://github.com/symfony/symfony/blob/56fa4dc6da1b50cd1d2be08d1ac3d704f2ade9a9/src/Symfony/Component/Cache/Traits/FilesystemTrait.php#L36

> So you might try to use the PdoAdapter together with SQLite for those systems that support it. We should check if performance is good for regular reads etc. as well but it might be even better than working with the filesystem as we use content digests. Content digests cause 2 lookups per read, so in the FilesystemAdapter that would be 2 fopen() handles. With SQLite this would end up with just one so it might be indeed faster.

I could try. But primarily I just want this to not happen within the HTTP request. When or how would Varnish prune its cache for example? (Does it ever? 🤔)

**Comment by Toflar** [comment]
Well, all the others are services. Deamons that run constantly. So they can just do things in the background themselves.

**Comment by aschempp** [comment]
> Well, you have to think of the HTTP Cache as its separate instance. A command runs within the app, the cache does not. So at the end of the day, it is an HTTP request anyway. No matter how you implement it 😊 So integrating it with the Messenger or anything like that won't work.

I don't think having a `PRUNE` HTTP request makes sense for a cache. I'm pretty sure Varnish (as an example) has some sort of command tools available, and they run on the same host as the cache server. It would make sense to have such tools as well. I agree with @Toflar though that _technically_, the cache does not run within our application, so a command of the application should not touch the cache. But maybe we can add some exception to the managed edition because we know how the cache works in that case? Or at least we can support the default case?

Btw, how does cache pruning work in the back end? Pretty sure we simply delete the cache directory?

**Comment by Toflar** [comment]
Pruning is not clearing. Clearing is fast because it kills the entire cache. Pruning means deleting only the expired items. This takes a very long time on the filesystem because it has to look into every single file to check if it is expired.
I'm working on an SQLite based solution which we dropped in the past because it did not support cache tagging. SQLite would be way faster in pruning because that's one query that can use an index. Let's see if we may use that soonish: https://github.com/symfony/symfony/pull/58296

**Comment by fritzmg** [comment]
> so a command of the application should not touch the cache

There already is

* `fos:httpcache:clear`
* `fos:httpcache:invalidate:path`
* `fos:httpcache:invalidate:regex`
* `fos:httpcache:invalidate:tag `
* `fos:httpcache:refresh:path`

fos just does not support prunable caches though (yet - that would be something I could contribute).

**Comment by fritzmg** [comment]
As discussed in the Contao call the correct solution would be to implement support for prunable caches in `friendsofsymfony/http-cache`.

Regardless, to alleviate the issue of pruning taking too long during the HTTP request @Toflar is working on a PR for Symfony for a `PdoTagAwareAdapter` that we can may be use in the future.

I also further alleviated the issue in the original instance by not caching certain, unimportant URLs via https://github.com/inspiredminds/contao-extended-cache-controls
